File: hypermixins-runtime/src/main/java/net/echo/hypermixins/agent/NameMangling.java
package net.echo.hypermixins.agent;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Mangled-name algorithm shared with the KSP processor. Must stay byte-for-byte identical with
 * {@code net.echo.hypermixins.processor.NameMangling} or compile-time-precomputed names will
 * not line up with runtime-produced synthetic helpers.
 */
public final class NameMangling {

    private NameMangling() {}

    public static String mangledOriginalName(String methodName, String descriptor) {
        return "__original$" + methodName + "$" + sha1Hex16(descriptor);
    }

    public static String dispatchName(String methodName, String descriptor) {
        return "__dispatch$" + methodName + "$" + sha1Hex16(descriptor);
    }

    public static String sha1Hex16(String input) {
        try {
            MessageDigest sha1 = MessageDigest.getInstance("SHA-1");
            byte[] hash = sha1.digest(input.getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder(16);
            for (int i = 0; i < 8; i++) hex.append(String.format("%02x", hash[i]));
            return hex.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

File: hypermixins-processor/src/main/kotlin/net/echo/hypermixins/processor/NameMangling.kt
package net.echo.hypermixins.processor

import java.nio.charset.StandardCharsets
import java.security.MessageDigest

/**
 * Mangled-name algorithm must stay byte-for-byte identical with
 * `MixinTransformer.mangledName` / `dispatchName` in the runtime module.
 * KSP precomputes these so the runtime never has to re-hash.
 */
internal object NameMangling {

    fun mangledOriginalName(methodName: String, descriptor: String): String =
        "__original$" + methodName + "$" + sha1Hex16(descriptor)

    fun dispatchName(methodName: String, descriptor: String): String =
        "__dispatch$" + methodName + "$" + sha1Hex16(descriptor)

    private fun sha1Hex16(input: String): String {
        val sha1 = MessageDigest.getInstance("SHA-1")
        val hash = sha1.digest(input.toByteArray(StandardCharsets.UTF_8))
        return buildString(16) {
            for (i in 0 until 8) append("%02x".format(hash[i]))
        }
    }
}
